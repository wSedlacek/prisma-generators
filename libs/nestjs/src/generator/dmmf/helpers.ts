// TODO: support not string args
export const modelAttributeRegex = /(@@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export const parseDocumentationAttributes = (
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: 'model' | 'field'
) => {
  const attributeRegex =
    expectedAttributeKind === 'model'
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return undefined;
  }

  return attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
};
