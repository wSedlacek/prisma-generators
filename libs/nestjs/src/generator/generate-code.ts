import { DMMF as PrismaDMMF } from '@prisma/client/runtime/dmmf-types';
import path from 'path';
import { Project } from 'ts-morph';

import { isDefined, saveSourceFile } from '../utils';
import { generateArgsTypeClassFromArgs } from './args-class';
import {
  argsFolderName,
  crudResolversFolderName,
  enumsFolderName,
  inputsFolderName,
  modelsFolderName,
  outputsFolderName,
  relationsResolversFolderName,
  resolversFolderName,
} from './config';
import { DmmfDocument } from './dmmf/dmmf-document';
import { DMMF } from './dmmf/types';
import { generateEnumFromDef } from './enum';
import { noop } from './helpers';
import {
  generateArgsBarrelFile,
  generateEnumsBarrelFile,
  generateIndexFile,
  generateInputsBarrelFile,
  generateModelsBarrelFile,
  generateOutputsBarrelFile,
  generateResolversBarrelFile,
} from './imports';
import { generateObjectTypeClassFromModel } from './model-type-class';
import { GenerateCodeOptions } from './options';
import { generateCrudResolverClassFromMapping } from './resolvers/full-crud';
import { generateRelationsResolverClassesFromModel } from './resolvers/relations';
import {
  generateInputTypeClassFromType,
  generateOutputTypeClassFromType,
} from './type-class';

export const generateCode = async (
  dmmf: PrismaDMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop
) => {
  const baseDirPath = options.outputDirPath;
  const project = new Project();
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map((model) => model.name);

  log('Transforming dmmfDocument...');
  const dmmfDocument = new DmmfDocument(dmmf, options);

  log('Generating enums...');
  const datamodelEnumNames = dmmfDocument.datamodel.enums.map(
    (enumDef) => enumDef.typeName
  );
  await Promise.all(
    dmmfDocument.datamodel.enums.map((enumDef) =>
      generateEnumFromDef(project, baseDirPath, enumDef)
    )
  );
  await Promise.all(
    dmmfDocument.schema.enums
      // skip enums from datamodel
      .filter((enumDef) => !datamodelEnumNames.includes(enumDef.typeName))
      .map((enumDef) => generateEnumFromDef(project, baseDirPath, enumDef))
  );
  const emittedEnumNames = [
    ...new Set([
      ...dmmfDocument.schema.enums.map((it) => it.typeName),
      ...dmmfDocument.datamodel.enums.map((it) => it.typeName),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);
  await saveSourceFile(enumsBarrelExportSourceFile);

  log('Generating models...');
  await Promise.all(
    dmmfDocument.datamodel.models.map((model) =>
      generateObjectTypeClassFromModel(
        project,
        baseDirPath,
        model,
        dmmfDocument,
        options
      )
    )
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmfDocument.datamodel.models.map((it) => it.typeName)
  );
  await saveSourceFile(modelsBarrelExportSourceFile);

  log('Generating output types...');
  const rootTypes = dmmfDocument.schema.outputTypes.filter((type) =>
    ['Query', 'Mutation'].includes(type.name)
  );
  const outputTypesToGenerate = dmmfDocument.schema.outputTypes.filter(
    // skip generating models and root resolvers
    (type) => !modelNames.includes(type.name) && !rootTypes.includes(type)
  );
  const outputTypesFieldsArgsToGenerate = outputTypesToGenerate
    .map((it) => it.fields)
    .reduce((a, b) => a.concat(b), [])
    .filter((it): it is Required<DMMF.OutputSchemaField> => !!it.argsTypeName);
  await Promise.all(
    outputTypesToGenerate.map((type) =>
      generateOutputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options
      )
    )
  );
  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map((it) => it.typeName),
    outputTypesFieldsArgsToGenerate.length > 0
  );
  await saveSourceFile(outputsBarrelExportSourceFile);

  if (outputTypesFieldsArgsToGenerate.length > 0) {
    log('Generating output types args...');
    await Promise.all(
      outputTypesFieldsArgsToGenerate.map(async (field) => {
        await generateArgsTypeClassFromArgs(
          project,
          path.resolve(resolversDirPath, outputsFolderName),
          field.args,
          field.argsTypeName,
          dmmfDocument,
          options,
          2
        );
      })
    );
    const outputsArgsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        outputsFolderName,
        argsFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateArgsBarrelFile(
      outputsArgsBarrelExportSourceFile,
      outputTypesFieldsArgsToGenerate
        .map((it) => it.argsTypeName)
        .filter(isDefined)
    );
    await saveSourceFile(outputsArgsBarrelExportSourceFile);
  }

  log('Generating input types...');
  await Promise.all(
    dmmfDocument.schema.inputTypes.map((type) =>
      generateInputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options
      )
    )
  );
  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmfDocument.schema.inputTypes.map((it) => it.typeName)
  );
  await saveSourceFile(inputsBarrelExportSourceFile);

  log('Generating relation resolvers...');
  const relationResolversData = await Promise.all(
    dmmfDocument.datamodel.models
      .filter((model) => model.fields.some((field) => field.relationName))
      .map(async (model) => {
        const outputType = dmmfDocument.schema.outputTypes.find(
          (type) => type.name === model.name
        );
        const mapping = dmmfDocument.mappings.find(
          (it) => it.model === model.name
        );

        if (!outputType || !mapping) {
          throw new Error('Could not determine resolver data');
        }

        return generateRelationsResolverClassesFromModel(
          project,
          baseDirPath,
          model,
          mapping,
          outputType,
          dmmfDocument,
          options
        );
      })
  );
  if (relationResolversData.length > 0) {
    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateResolversBarrelFile(
      'relations',
      relationResolversBarrelExportSourceFile,
      relationResolversData
    );
    await saveSourceFile(relationResolversBarrelExportSourceFile);
  }

  log('Generating crud resolvers...');
  const crudResolversData = await Promise.all(
    dmmfDocument.mappings.map(async (mapping) => {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      );

      if (!model) {
        throw new Error('Could not determine resolver data');
      }

      return generateCrudResolverClassFromMapping(
        project,
        baseDirPath,
        mapping,
        model,
        modelNames,
        options,
        dmmfDocument
      );
    })
  );
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateResolversBarrelFile(
    'crud',
    crudResolversBarrelExportSourceFile,
    crudResolversData
  );
  await saveSourceFile(crudResolversBarrelExportSourceFile);

  log('Generating index file');
  const indexSourceFile = project.createSourceFile(
    `${baseDirPath}/index.ts`,
    undefined,
    { overwrite: true }
  );
  generateIndexFile(indexSourceFile, relationResolversData.length > 0);
  await saveSourceFile(indexSourceFile);
};
