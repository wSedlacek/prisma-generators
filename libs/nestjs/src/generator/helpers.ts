import { DmmfDocument } from './dmmf/dmmf-document';
import { fieldAttributeRegex, modelAttributeRegex } from './dmmf/helpers';
import { DMMF } from './dmmf/types';

export const noop = () => {};

export const getFieldTSType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  isInputType: boolean,
  modelName?: string,
  typeName?: string
) => {
  let TSType: string;
  if (typeInfo.kind === 'scalar') {
    TSType = mapScalarToTSType(typeInfo.type, isInputType);
  } else if (typeInfo.kind === 'object') {
    TSType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else if (typeInfo.kind === 'enum') {
    TSType = `keyof typeof ${typeInfo.type}`;
  } else {
    throw new Error(`Unsupported field type kind: ${typeInfo.kind}`);
  }
  if (typeInfo.isList) {
    if (TSType.includes(' ')) {
      TSType = `Array<${TSType}>`;
    } else {
      TSType += '[]';
    }
  }
  if (!typeInfo.isRequired) {
    // FIXME: use properly null for output and undefined for input
    // TSType += ' | undefined';
    TSType += ' | null | undefined';
  }

  return TSType;
};

export const mapScalarToTSType = (scalar: string, isInputType: boolean) => {
  switch (scalar) {
    case 'ID':
    case 'UUID': {
      return 'string';
    }
    case 'String': {
      return 'string';
    }
    case 'Boolean': {
      return 'boolean';
    }
    case 'DateTime': {
      return 'Date';
    }
    case 'Int':
    case 'Float': {
      return 'number';
    }
    case 'Json':
      return isInputType ? 'InputJsonValue' : 'JsonValue';
    default:
      throw new Error(`Unrecognized scalar type: ${scalar}`);
  }
};

export const getGraphQLType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  modelName?: string,
  typeName?: string
) => {
  let GraphQLType: string;
  if (typeInfo.kind === 'scalar') {
    GraphQLType = mapScalarToType(typeInfo.type);
  } else if (typeInfo.kind === 'object') {
    GraphQLType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else {
    GraphQLType = typeInfo.type;
  }
  if (typeInfo.isList) {
    GraphQLType = `[${GraphQLType}]`;
  }

  return GraphQLType;
};

export const mapScalarToType = (scalar: string) => {
  switch (scalar) {
    case 'DateTime': {
      return 'Date';
    }
    // TODO: use proper uuid graphql scalar
    case 'UUID': {
      return 'String';
    }
    case 'Boolean':
    case 'String': {
      return scalar;
    }
    case 'ID':
    case 'Int':
    case 'Float': {
      return `${scalar}`;
    }
    case 'Json': {
      return `GraphQLJSON`;
    }
    default: {
      throw new Error(`Unrecognized scalar type: ${scalar}`);
    }
  }
};

export const camelCase = (str: string) => {
  return str[0].toLowerCase() + str.slice(1);
};

export const pascalCase = (str: string): string => {
  return str[0].toUpperCase() + str.slice(1);
};

const getInputKeywordPhrasePosition = (inputTypeName: string) => {
  const inputParseResult = [
    'Create',
    'OrderBy',
    'Update',
    'Upsert',
    'ScalarWhere',
    'Where',
    'ListRelationFilter',
    'RelationFilter',
    'Filter',
  ]
    .map((inputKeyword) => inputTypeName.search(inputKeyword))
    .filter((position) => position >= 0);

  if (inputParseResult.length === 0) {
    return undefined;
  }

  return inputParseResult[0];
};

export const getModelNameFromInputType = (inputTypeName: string) => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(inputTypeName);
  if (!keywordPhrasePosition) {
    return undefined;
  }

  return inputTypeName.slice(0, keywordPhrasePosition);
};

export const getInputTypeName = (
  originalInputName: string,
  dmmfDocument: DmmfDocument
): string => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(
    originalInputName
  );
  if (!keywordPhrasePosition) {
    return originalInputName;
  }

  const modelName = originalInputName.slice(0, keywordPhrasePosition);
  const typeNameRest = originalInputName.slice(keywordPhrasePosition);
  const modelTypeName = dmmfDocument.getModelTypeName(modelName);
  if (!modelTypeName) {
    return originalInputName;
  }

  return `${modelTypeName}${typeNameRest}`;
};

export const cleanDocsString = (
  documentation: string | undefined
): string | undefined => {
  if (!documentation) {
    return undefined;
  }
  let cleanedDocs = documentation;
  cleanedDocs = cleanedDocs.replace(modelAttributeRegex, '');
  cleanedDocs = cleanedDocs.replace(fieldAttributeRegex, '');
  cleanedDocs = cleanedDocs.split('"').join('\\"');
  cleanedDocs = cleanedDocs.split('\r').join('');
  cleanedDocs = cleanedDocs.split('\\r').join('');
  cleanedDocs = cleanedDocs.split('\n').join('');
  cleanedDocs = cleanedDocs.split('\\n').join('');

  return cleanedDocs;
};

export const toUnixPath = (maybeWindowsPath: string) => {
  return maybeWindowsPath.split('\\').join('/');
};
