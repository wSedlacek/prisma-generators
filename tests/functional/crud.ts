import "reflect-metadata";
import { promises as fs } from "fs";
import { buildSchema } from "type-graphql";
import { graphql, GraphQLSchema } from "graphql";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { generateCodeFromSchema } from "../helpers/generate-code";

describe("crud resolvers execution", () => {
  let outputDirPath: string;
  let graphQLSchema: GraphQLSchema;

  describe("with default settings", () => {
    beforeAll(async () => {
      outputDirPath = generateArtifactsDirPath("functional-crud");
      await fs.mkdir(outputDirPath, { recursive: true });
      const prismaSchema = /* prisma */ `
        model User {
          intIdField          Int     @id @default(autoincrement())
          uniqueStringField   String  @unique
          optionalStringField String?
          dateField           DateTime
        }
      `;
      await generateCodeFromSchema(prismaSchema, { outputDirPath });
      const { UserCrudResolver } = require(outputDirPath +
        "/resolvers/crud/User/UserCrudResolver.ts");

      graphQLSchema = await buildSchema({
        resolvers: [UserCrudResolver],
        validate: false,
      });
    });

    it("should properly call PrismaClient on `findOne` action", async () => {
      const document = /* graphql */ `
        query {
          user(where: { uniqueStringField: "uniqueValue" }) {
            intIdField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          findOne: jest.fn().mockResolvedValue({
            intIdField: 1,
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          }),
        },
      };

      const { data, errors } = await graphql(graphQLSchema, document, null, {
        prisma: prismaMock,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("user mocked response");
      expect(prismaMock.user.findOne.mock.calls).toMatchSnapshot(
        "findOneUser call args",
      );
    });

    it("should properly call PrismaClient on `findMany` action", async () => {
      const document = /* graphql */ `
        query {
          users(
            take: 1
            skip: 1
            orderBy: { intIdField: desc }
            where: { dateField: { lte: "2019-12-31T19:16:02.572Z" } }
          ) {
            intIdField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          findMany: jest.fn().mockResolvedValue([
            {
              intIdField: 1,
              dateField: new Date("2019-12-31T14:16:02.572Z"),
            },
          ]),
        },
      };

      const { data, errors } = await graphql(graphQLSchema, document, null, {
        prisma: prismaMock,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("users mocked response");
      expect(prismaMock.user.findMany.mock.calls).toMatchSnapshot(
        "findManyUser call args",
      );
    });

    it("should properly call PrismaClient on `create` action", async () => {
      const document = /* graphql */ `
        mutation {
          createUser(
            data: {
              uniqueStringField: "unique"
              optionalStringField: "optional"
              dateField: "2019-12-31T14:16:02.572Z"
            }
          ) {
            intIdField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          create: jest.fn().mockResolvedValue({
            intIdField: 1,
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          }),
        },
      };

      const { data, errors } = await graphql(graphQLSchema, document, null, {
        prisma: prismaMock,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("createUser mocked response");
      expect(prismaMock.user.create.mock.calls).toMatchSnapshot(
        "createUser call args",
      );
    });

    it("should properly call PrismaClient on `delete` action", async () => {
      const document = /* graphql */ `
        mutation {
          deleteUser(
            where: {
              uniqueStringField: "unique"
            }
          ) {
            intIdField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          delete: jest.fn().mockResolvedValue({
            intIdField: 1,
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          }),
        },
      };

      const { data, errors } = await graphql(graphQLSchema, document, null, {
        prisma: prismaMock,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("deleteUser mocked response");
      expect(prismaMock.user.delete.mock.calls).toMatchSnapshot(
        "deleteUser call args",
      );
    });

    it("should properly call PrismaClient on `update` action", async () => {
      const document = /* graphql */ `
        mutation {
          updateUser(
            data: {
              dateField: "2019-12-31T14:16:02.572Z",
            }
            where: {
              uniqueStringField: "unique"
            }
          ) {
            intIdField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          update: jest.fn().mockResolvedValue({
            intIdField: 1,
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          }),
        },
      };

      const { data, errors } = await graphql(graphQLSchema, document, null, {
        prisma: prismaMock,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("updateUser mocked response");
      expect(prismaMock.user.update.mock.calls).toMatchSnapshot(
        "updateUser call args",
      );
    });

    it("should properly call PrismaClient on `updateMany` action", async () => {
      const document = /* graphql */ `
        mutation {
          updateManyUser(
            data: {
              optionalStringField: null,
            }
            where: {
              dateField: { lte: "2019-12-31T19:16:02.572Z" }
            }
          ) {
            count
          }
        }
      `;
      const prismaMock = {
        user: {
          updateMany: jest.fn().mockResolvedValue({
            count: 3,
          }),
        },
      };

      const { data, errors } = await graphql(graphQLSchema, document, null, {
        prisma: prismaMock,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("updateManyUser mocked response");
      expect(prismaMock.user.updateMany.mock.calls).toMatchSnapshot(
        "updateManyUser call args",
      );
    });

    it("should properly call PrismaClient on `deleteMany` action", async () => {
      const document = /* graphql */ `
        mutation {
          deleteManyUser(
            where: {
              dateField: { lte: "2019-12-31T19:16:02.572Z" }
            }
          ) {
            count
          }
        }
      `;
      const prismaMock = {
        user: {
          deleteMany: jest.fn().mockResolvedValue({
            count: 3,
          }),
        },
      };

      const { data, errors } = await graphql(graphQLSchema, document, null, {
        prisma: prismaMock,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("deleteManyUser mocked response");
      expect(prismaMock.user.deleteMany.mock.calls).toMatchSnapshot(
        "deleteManyUser call args",
      );
    });

    it("should properly call PrismaClient on `upsert` action", async () => {
      const document = /* graphql */ `
        mutation {
          upsertUser(
            where: {
              uniqueStringField: "unique"
            }
            create: {
              uniqueStringField: "unique"
              optionalStringField: "optional"
              dateField: "2019-12-31T14:16:02.572Z"
            }
            update: {
              optionalStringField: null,
            }
          ) {
            intIdField
            uniqueStringField
            optionalStringField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          upsert: jest.fn().mockResolvedValue({
            intIdField: 1,
            uniqueStringField: "unique",
            optionalStringField: "optional",
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          }),
        },
      };

      const { data, errors } = await graphql(graphQLSchema, document, null, {
        prisma: prismaMock,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("upsertUser mocked response");
      expect(prismaMock.user.upsert.mock.calls).toMatchSnapshot(
        "upsertUser call args",
      );
    });

    it("should properly call PrismaClient on `aggregate` action with simple count field", async () => {
      const document = /* graphql */ `
        query {
          aggregateUser(
            take: 1
            skip: 1
            orderBy: { intIdField: desc }
            where: { dateField: { lte: "2019-12-31T19:16:02.572Z" } }
          ) {
            count
          }
        }
      `;
      const prismaMock = {
        user: {
          aggregate: jest.fn().mockResolvedValue({ count: 5 }),
        },
      };

      const { data, errors } = await graphql(graphQLSchema, document, null, {
        prisma: prismaMock,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("aggregateUserCount mocked response");
      expect(prismaMock.user.aggregate.mock.calls).toMatchSnapshot(
        "user.aggregate call args",
      );
    });
  });

  describe("with experimental aggregations api enabled", () => {
    beforeAll(async () => {
      outputDirPath = generateArtifactsDirPath("functional-crud");
      await fs.mkdir(outputDirPath, { recursive: true });
      const prismaSchema = /* prisma */ `
        model User {
          idField     Int  @id @default(autoincrement())
          intField    Int
          floatField  Int
        }
      `;
      await generateCodeFromSchema(prismaSchema, {
        outputDirPath,
        enabledPreviewFeatures: ["aggregations"],
      });
      const { UserCrudResolver } = require(outputDirPath +
        "/resolvers/crud/User/UserCrudResolver.ts");

      graphQLSchema = await buildSchema({
        resolvers: [UserCrudResolver],
        validate: false,
      });
    });

    it("should properly call PrismaClient on `aggregate` action with advanced operations", async () => {
      const document = /* graphql */ `
        query {
          aggregateUser(
            take: 1
            skip: 1
            orderBy: { intField: desc }
            where: { floatField: { lte: 50 } }
          ) {
            __typename
            count
            min {
              __typename
              intField
              floatField
            }
            max {
              __typename
              intField
              floatField
            }
            sum {
              __typename
              intField
              floatField
            }
            avg {
              __typename
              intField
              floatField
            }
          }
        }
      `;
      const prismaMock = {
        user: {
          aggregate: jest.fn().mockResolvedValue({
            count: 2,
            min: {
              intField: 0,
              floatField: 0,
            },
            max: {
              intField: 10,
              floatField: 10,
            },
            sum: {
              intField: 10,
              floatField: 10,
            },
            avg: {
              intField: 5,
              floatField: 5,
            },
          }),
        },
      };

      const { data, errors } = await graphql(graphQLSchema, document, null, {
        prisma: prismaMock,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("aggregateUser mocked response");
      expect(prismaMock.user.aggregate.mock.calls).toMatchSnapshot(
        "user.aggregate call args",
      );
    });
  });
});
