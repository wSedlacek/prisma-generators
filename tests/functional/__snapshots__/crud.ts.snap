// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`aggregate\` action with simple count field: aggregateUserCount mocked response 1`] = `
Object {
  "aggregateUser": Object {
    "count": 5,
  },
}
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`aggregate\` action with simple count field: user.aggregate call args 1`] = `
Array [
  Array [
    Object {
      "count": true,
      "orderBy": UserOrderByInput {
        "intIdField": "desc",
      },
      "skip": 1,
      "take": 1,
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`create\` action: createUser call args 1`] = `
Array [
  Array [
    CreateUserArgs {
      "data": UserCreateInput {
        "dateField": 2019-12-31T14:16:02.572Z,
        "optionalStringField": "optional",
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`create\` action: createUser mocked response 1`] = `
Object {
  "createUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`delete\` action: deleteUser call args 1`] = `
Array [
  Array [
    DeleteUserArgs {
      "where": UserWhereUniqueInput {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`delete\` action: deleteUser mocked response 1`] = `
Object {
  "deleteUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`deleteMany\` action: deleteManyUser call args 1`] = `
Array [
  Array [
    DeleteManyUserArgs {
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`deleteMany\` action: deleteManyUser mocked response 1`] = `
Object {
  "deleteManyUser": Object {
    "count": 3,
  },
}
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`findMany\` action: findManyUser call args 1`] = `
Array [
  Array [
    FindManyUserArgs {
      "orderBy": UserOrderByInput {
        "intIdField": "desc",
      },
      "skip": 1,
      "take": 1,
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`findMany\` action: users mocked response 1`] = `
Object {
  "users": Array [
    Object {
      "dateField": "2019-12-31T14:16:02.572Z",
      "intIdField": 1,
    },
  ],
}
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`findOne\` action: findOneUser call args 1`] = `
Array [
  Array [
    FindOneUserArgs {
      "where": UserWhereUniqueInput {
        "uniqueStringField": "uniqueValue",
      },
    },
  ],
]
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`findOne\` action: user mocked response 1`] = `
Object {
  "user": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`update\` action: updateUser call args 1`] = `
Array [
  Array [
    UpdateUserArgs {
      "data": UserUpdateInput {
        "dateField": 2019-12-31T14:16:02.572Z,
      },
      "where": UserWhereUniqueInput {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`update\` action: updateUser mocked response 1`] = `
Object {
  "updateUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`updateMany\` action: updateManyUser call args 1`] = `
Array [
  Array [
    UpdateManyUserArgs {
      "data": UserUpdateManyMutationInput {
        "optionalStringField": null,
      },
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`updateMany\` action: updateManyUser mocked response 1`] = `
Object {
  "updateManyUser": Object {
    "count": 3,
  },
}
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`upsert\` action: upsertUser call args 1`] = `
Array [
  Array [
    UpsertUserArgs {
      "create": UserCreateInput {
        "dateField": 2019-12-31T14:16:02.572Z,
        "optionalStringField": "optional",
        "uniqueStringField": "unique",
      },
      "update": UserUpdateInput {
        "optionalStringField": null,
      },
      "where": UserWhereUniqueInput {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution with default settings should properly call PrismaClient on \`upsert\` action: upsertUser mocked response 1`] = `
Object {
  "upsertUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
    "optionalStringField": "optional",
    "uniqueStringField": "unique",
  },
}
`;

exports[`crud resolvers execution with experimental aggregations api enabled should properly call PrismaClient on \`aggregate\` action with advanced operations: aggregateUser mocked response 1`] = `
Object {
  "aggregateUser": Object {
    "__typename": "AggregateUser",
    "avg": Object {
      "__typename": "UserAvgAggregateOutputType",
      "floatField": 5,
      "intField": 5,
    },
    "count": 2,
    "max": Object {
      "__typename": "UserMaxAggregateOutputType",
      "floatField": 10,
      "intField": 10,
    },
    "min": Object {
      "__typename": "UserMinAggregateOutputType",
      "floatField": 0,
      "intField": 0,
    },
    "sum": Object {
      "__typename": "UserSumAggregateOutputType",
      "floatField": 10,
      "intField": 10,
    },
  },
}
`;

exports[`crud resolvers execution with experimental aggregations api enabled should properly call PrismaClient on \`aggregate\` action with advanced operations: user.aggregate call args 1`] = `
Array [
  Array [
    Object {
      "avg": Object {
        "floatField": true,
        "intField": true,
      },
      "count": true,
      "max": Object {
        "floatField": true,
        "intField": true,
      },
      "min": Object {
        "floatField": true,
        "intField": true,
      },
      "orderBy": UserOrderByInput {
        "intField": "desc",
      },
      "skip": 1,
      "sum": Object {
        "floatField": true,
        "intField": true,
      },
      "take": 1,
      "where": UserWhereInput {
        "floatField": IntFilter {
          "lte": 50,
        },
      },
    },
  ],
]
`;
